File: golabs/godata/go-gumball-riak/gumball-riak/src/github.com/basho/taste-of-riak/java/Ch03-Msgy-Schema/src/com/basho/msgy/Models/Msg.java
package com.basho.msgy.Models;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;

public class Msg {
    public String Sender;
    public String Recipient;
    public String Created;
    public String Text;

    public static Msg createNew(String sender, String recipient, String text) {
        Msg msg = new Msg();
        msg.Sender = sender;
        msg.Recipient = recipient;
        msg.Text = text;
        msg.Created = GetCurrentISO8601Timestamp();
        return msg;
    }

    private static String GetCurrentISO8601Timestamp() {
        TimeZone tz = TimeZone.getTimeZone("UTC");
        // Java Dates don't have microsecond resolution :(
        // Pad out to microseconds to match other examples.
        DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'000'");
        df.setTimeZone(tz);
        return df.format(new Date());
    }
}

File: golabs/gocloud/godata/go-gumball-riak/gumball-riak/src/github.com/basho/taste-of-riak/go/ch03/models/msg.go
package models

import (
	"fmt"
	"time"

	util "github.com/basho/taste-of-riak/go/util"
)

type Msg struct {
	modelImpl
	Sender    string
	Recipient string
	Text      string
	Created   time.Time
}

func NewMsg(sender, recipient, text string) *Msg {
	m := &Msg{
		Sender:    sender,
		Recipient: recipient,
		Text:      text,
		Created:   time.Now(),
	}
	m.SetId(m.GetId())
	return m
}

func (m *Msg) GetId() string {
	return fmt.Sprintf("%s_%v", m.Sender, util.Iso8601(m.Created))
}
